src/descriptor/checksum.rs

```diff
@@ -2,177 +2,180 @@
 
 //! Descriptor checksum
 //!
-//! This module contains a re-implementation of the function used by Bitcoin Core to calculate the
-//! checksum of a descriptor. The checksum algorithm is specified in [BIP-380].
+//! Descriptors are checksummed using [BCH] codes, the algorithm is defined in [BIP-380].
 //!
 //! [BIP-380]: <https://github.com/bitcoin/bips/blob/master/bip-0380.mediawiki>
+//! [BCH]: <https://en.wikipedia.org/wiki/BCH_code>
 
-use core::fmt;
-use core::iter::FromIterator;
+use core::convert::TryFrom;
+use core::str::Chars;
+
+use bech32::primitives::checksum::{self, PackedFe32};
+use bech32::{Checksum, Fe32};
 
 pub use crate::expression::VALID_CHARS;
 use crate::prelude::*;
 use crate::Error;
 
-const CHECKSUM_CHARSET: &[u8] = b"qpzry9x8gf2tvdw0s3jn54khce6mua7l";
-
-const CHECKSUM_LENGTH: usize = 8;
-
-fn poly_mod(mut c: u64, val: u64) -> u64 {
-    let c0 = c >> 35;
-
-    c = ((c & 0x7ffffffff) << 5) ^ val;
-    if c0 & 1 > 0 {
-        c ^= 0xf5dee51989
-    };
-    if c0 & 2 > 0 {
-        c ^= 0xa9fdca3312
-    };
-    if c0 & 4 > 0 {
-        c ^= 0x1bab10e32d
-    };
-    if c0 & 8 > 0 {
-        c ^= 0x3706b1677a
-    };
-    if c0 & 16 > 0 {
-        c ^= 0x644d626ffd
-    };
-
-    c
-}
-
-/// Compute the checksum of a descriptor.
-///
-/// Note that this function does not check if the descriptor string is
-/// syntactically correct or not. This only computes the checksum.
-pub fn desc_checksum(desc: &str) -> Result<String, Error> {
-    let mut eng = Engine::new();
-    eng.input(desc)?;
-    Ok(eng.checksum())
+/// Iterator adapter that yields the input descriptor string characters,
+/// followed by the # character, followed by the checksum characters.
+pub struct Checksummed<'a> {
+    iter: Chars<'a>, // Valid descriptor characters.
+    checksum_engine: checksum::Engine<DescriptorChecksum>,
+    checksum_remaining: usize,
+    cls: u64,
+    clscount: u64,
+    yielded_hash: bool, // `true` if we have yielded the `#` character already.
 }
 
-/// Helper function for `FromStr` for various descriptor types.
-///
-/// Checks and verifies the checksum if it is present and returns the descriptor
-/// string without the checksum.
-pub(super) fn verify_checksum(s: &str) -> Result<&str, Error> {
-    for ch in s.as_bytes() {
-        if *ch < 20 || *ch > 127 {
-            return Err(Error::Unprintable(*ch));
-        }
+impl<'a> Checksummed<'a> {
+    /// Creates a new checksummed iterator which adapts a descriptor string
+    /// iterator by appending a checksum.
+    ///
+    /// # Errors
+    ///
+    /// If descriptor contains invalid characters as specified by [BIP-380].
+    ///
+    /// [BIP-380]: <https://github.com/bitcoin/bips/blob/master/bip-0380.mediawiki>
+    #[inline]
+    pub fn new(descriptor: &'a str) -> Result<Checksummed<'a>, Error> {
```

Comment:
In 22ee854c61b13d15fb7d0b11daf2a87604f45ba7:

Rather than taking a string this should take a `D: Display` and a `fmt::Formatter`.